File: ts/2022/day14.ts
import { main } from 'lib/advent';
import { Dir, move, Point, PointMap } from 'lib/coords';
import { iter } from 'lib/iter';
import { lines, ValuesOf } from 'lib/util';

const EMITTER = { x: 500, y: 0 };
const Cell = {
  ROCK: '#',
  SAND: 'o',
} as const;
type Cell = ValuesOf<typeof Cell>;
type Cave = PointMap<Cell>;

function parse(lines: string[]) {
  return lines.map((line) => {
    return line.split(' -> ').map((s) => {
      const [x, y] = s.split(',').map(Number);
      return { x, y };
    });
  });
}

function makeCave(segments: Point[][]): Cave {
  const cave = new PointMap<Cell>();
  for (const points of segments) {
    const cur = { ...points[0] };
    cave.set(cur, Cell.ROCK);
    for (const point of points.slice(1)) {
      while (cur.x !== point.x) {
        cur.x += Math.sign(point.x - cur.x);
        cave.set(cur, Cell.ROCK);
      }
      while (cur.y !== point.y) {
        cur.y += Math.sign(point.y - cur.y);
        cave.set(cur, Cell.ROCK);
      }
    }
  }
  return cave;
}

function overflowCave(cave: Cave): Cave {
  const bottom = iter(cave.keys())
    .map((p) => p.y)
    .max();

  for (;;) {
    let sand = move(EMITTER, Dir.Down);
    for (;;) {
      const down = move(sand, Dir.Down);
      if (down.y > bottom) return cave;
      const fall = [down, move(down, Dir.Left), move(down, Dir.Right)].find(
        (p) => !cave.has(p)
      );
      if (fall) sand = fall;
      else {
        cave.set(sand, Cell.SAND);
        break;
      }
    }
  }
}

function fillCave(cave: Cave): Cave {
  const bottom =
    iter(cave.keys())
      .map((p) => p.y)
      .max() + 2;

  while (!cave.has(EMITTER)) {
    let sand = EMITTER;
    for (;;) {
      const down = move(sand, Dir.Down);
      const fall = [down, move(down, Dir.Left), move(down, Dir.Right)].find(
        (p) => p.y < bottom && !cave.has(p)
      );
      if (fall) sand = fall;
      else {
        cave.set(sand, Cell.SAND);
        break;
      }
    }
  }
  return cave;
}

function countSand(cave: Cave): number {
  return cave.values().filter((v) => v === Cell.SAND).length;
}

main(
  (s) => countSand(overflowCave(makeCave(parse(lines(s))))),
  (s) => countSand(fillCave(makeCave(parse(lines(s)))))
);

File: py/2022/day14.py
from itertools import batched

from aoc import main
from aoc.coords import Dir, Dir8, Point, addp, line_between
from aoc.parse import all_numbers, line_parser
from aoc.util import sliding_window

EMITTER = (500, 0)


@line_parser
def parse(line: str):
    points = [(a, b) for a, b in batched(all_numbers(line), 2)]
    return {p for a, b in sliding_window(points, 2) for p in line_between(a, b)}


def drop_sand(filled: set[Point], bottom: int, solid_bottom=False) -> Point:
    cur = EMITTER
    while True:
        if cur[1] == bottom - (1 if solid_bottom else 0):
            break
        elif (
            (p := addp(cur, Dir.S)) not in filled
            or (p := addp(cur, Dir8.SW)) not in filled
            or (p := addp(cur, Dir8.SE)) not in filled
        ):
            pass
        else:
            break
        cur = p
    return cur


def sand_at_rest(s: str) -> int:
    rocks: set[Point] = set.union(*parse(s))
    bottom = max(p[1] for p in rocks)

    sand: set[Point] = set()
    while True:
        grain = drop_sand(rocks | sand, bottom)
        if grain[1] >= bottom:
            break
        sand.add(grain)
    return len(sand)


def sand_until_blocked(s: str) -> int:
    rocks: set[Point] = set.union(*parse(s))
    bottom = max(p[1] for p in rocks) + 2

    sand: set[Point] = set()
    while True:
        grain = drop_sand(rocks, bottom, solid_bottom=True)
        sand.add(grain)
        rocks.add(grain)
        if grain == EMITTER:
            break
    return len(sand)


if __name__ == "__main__":
    main(sand_at_rest, sand_until_blocked, isolate=0)
